edgeless_api/src/function_instance/mod.rs

```diff
@@ -65,7 +82,7 @@ pub struct UpdateFunctionLinksRequest {
 
 #[async_trait::async_trait]
 pub trait FunctionInstanceAPI: FunctionInstanceAPIClone + Sync + Send {
-    async fn start(&mut self, spawn_request: SpawnFunctionRequest) -> anyhow::Result<InstanceId>;
+    async fn start(&mut self, spawn_request: SpawnFunctionRequest) -> anyhow::Result<SpawnFunctionResponse>;
     async fn stop(&mut self, id: InstanceId) -> anyhow::Result<()>;
     async fn update_links(&mut self, update: UpdateFunctionLinksRequest) -> anyhow::Result<()>;
```

Comment:
Given this is mostly a start request this will need similar semantics.

Reply:
Do you mean for the update_links()?

If yes, then I am not sure I can agree:

Starting a new instance is something that may need _a lot_ of resources and might have constraints. There are many reasons why one should refuse (not enough memory available, cores overloaded, annotation specifies the need for a GPU but it is already used/locked by another process, ...).

The update_links does not require more resources because it is a mere update of a forwarding table, which should not fail unless exceptional circumstances happen (I don't know, not enough memory to add few bytes needed to the data structure?).

IMO, we _should_ use fail-semantics for start-like methods, which may be rejected in non-exceptional circumstances, while we do not need to require fail-semantics for those that may only fail in exceptional cases.

Reply:
Yes I mean update links. One can still think of some scenarios that might lead to a rejection (e.g. when violating some security policy by trying to link to some other workflow).

Reply:
I see two consistent ways of making things.

A. We implement the fail-semantics on 100% of the calls, e.g., read-only methods. That's fine, since _something can always go wrong in life_.

B. We implement the fail-semantics only on methods that may be willingly *rejected*.

I don't like A at this stage because things are moving and there will be maintenance cost to be paid.

If we go for B, as I recommend for now, then I think that `update_links` does not fall into the category of methods for which the node may reject that.

The violation of a security policy is something which has not been defined, as of today. When this will be a thing, then `update_links` will fall into the right category add we will add the fail-semantics accordingly.

I think it's more efficient to proceed on a "let's do only things that makes sense today" way and not to overdesign because our resources are very limited.